### ISSUE TITLE
Harden services.sh against gcloud errors

### ISSUE THREAD

**Issue by satnam6502** [GoogleCloudPlatform/kubernetes#3951]
The `services.sh` e2e test fails when a gcloud command it executes returns a `Backend error`. We should harden this test against such failures e.g. by retyring gcloud commands that fail. It would be great if someone who knows this test could take a look and find out where the underlying gcloud commands are (they are not directly in this file). This is causing our e2e tests to fail so I am assigning this a P1 priority. Please can one of @brendandburns @smreed @zmerlynn @thockin take this?

Example output from Jenkins GCE run #1667:
```
not ok 4 - services.sh
  ---
  duration_ms: 111.725
  stdout: |
    Found e2e-test-jenkins-minion-1 at 104.154.43.250
    Found e2e-test-jenkins-minion-2 at 130.211.133.125
    Starting service 'service1' on port 80 with 3 replicas
    service1
    service1
    Starting service 'service2' on port 80 with 3 replicas
    service2
    service2
    Querying pods in service1
     a3777d3e-a7f9-11e4-9948-42010af0c23d
     a377fabb-a7f9-11e4-9948-42010af0c23d
     a3781adc-a7f9-11e4-9948-42010af0c23d
    Waiting for 3 pods to become 'running'
    Querying pods in service2
     a9a0e74c-a7f9-11e4-9948-42010af0c23d
     a9a15112-a7f9-11e4-9948-42010af0c23d
     a9a19402-a7f9-11e4-9948-42010af0c23d
    Waiting for 3 pods to become 'running'
    Test 1꞉ Prove that the service portal is alive.
    Verifying the portals from the host
    Checking if a3777d3e-a7f9-11e4-9948-42010af0c23d a377fabb-a7f9-11e4-9948-42010af0c23d a3781adc-a7f9-11e4-9948-42010af0c23d  == a3777d3e-a7f9-11e4-9948-42010af0c23d a377fabb-a7f9-11e4-9948-42010af0c23d a3781adc-a7f9-11e4-9948-42010af0c23d 
    Checking if a3777d3e-a7f9-11e4-9948-42010af0c23d a377fabb-a7f9-11e4-9948-42010af0c23d a3781adc-a7f9-11e4-9948-42010af0c23d  == a3777d3e-a7f9-11e4-9948-42010af0c23d a377fabb-a7f9-11e4-9948-42010af0c23d a3781adc-a7f9-11e4-9948-42010af0c23d 
    Checking if a3777d3e-a7f9-11e4-9948-42010af0c23d a377fabb-a7f9-11e4-9948-42010af0c23d a3781adc-a7f9-11e4-9948-42010af0c23d  == a3777d3e-a7f9-11e4-9948-42010af0c23d a377fabb-a7f9-11e4-9948-42010af0c23d a3781adc-a7f9-11e4-9948-42010af0c23d 
    Checking if a9a0e74c-a7f9-11e4-9948-42010af0c23d a9a15112-a7f9-11e4-9948-42010af0c23d a9a19402-a7f9-11e4-9948-42010af0c23d  == a9a0e74c-a7f9-11e4-9948-42010af0c23d a9a15112-a7f9-11e4-9948-42010af0c23d a9a19402-a7f9-11e4-9948-42010af0c23d 
    Verifying the portals from a container
    
  stderr: |
    Project꞉ kubernetes-jenkins
    Zone꞉ us-central1-f
    Project꞉ kubernetes-jenkins
    Zone꞉ us-central1-f
    Running꞉ /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/cluster/../cluster/gce/../../platforms/linux/amd64/kubectl --match-server-version create -f -
    Running꞉ /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/cluster/../cluster/gce/../../platforms/linux/amd64/kubectl --match-server-version create -f -
    Running꞉ /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/cluster/../cluster/gce/../../platforms/linux/amd64/kubectl --match-server-version create -f -
    Running꞉ /jenkins-master-data/jobs/kubernetes-e2e-gce/workspace/kubernetes/cluster/../cluster/gce/../../platforms/linux/amd64/kubectl --match-server-version create -f -
    Project꞉ kubernetes-jenkins
    Zone꞉ us-central1-f
    Project꞉ kubernetes-jenkins
    Zone꞉ us-central1-f
    Project꞉ kubernetes-jenkins
    Zone꞉ us-central1-f
    Project꞉ kubernetes-jenkins
    Zone꞉ us-central1-f
    Project꞉ kubernetes-jenkins
    Zone꞉ us-central1-f
    Project꞉ kubernetes-jenkins
    Zone꞉ us-central1-f
    Project꞉ kubernetes-jenkins
    Zone꞉ us-central1-f
    Project꞉ kubernetes-jenkins
    Zone꞉ us-central1-f
    Project꞉ kubernetes-jenkins
    Zone꞉ us-central1-f
    Project꞉ kubernetes-jenkins
    Zone꞉ us-central1-f
    Project꞉ kubernetes-jenkins
    Zone꞉ us-central1-f
    Project꞉ kubernetes-jenkins
    Zone꞉ us-central1-f
    ERROR꞉ (gcloud.compute.ssh) Could not fetch project resource꞉
     - Backend Error
    
    testing service1 portal from container failed
```

**Comment by thockin** [comment]
I'm not sure what it means to harden against that failure - does 'gcloud' return a particular exit code?

The calls to gcloud come from ssh-to-node()

**Comment by thockin** [comment]
Sadly, gcloud does not make it possible to know whether gcloud failed or whether it made an SSH connection and the command that was run returned 1.

Suggestions, o test wizards?

**Comment by satnam6502** [comment]
If we know the command is idempotent then surely we could add a retry loop to:
```
# SSH to a node by name ($1) and run a command ($2).
function ssh-to-node {
  local node="$1"
  local cmd="$2"
  gcloud compute ssh --ssh-flag="-o LogLevel=quiet" --project "${PROJECT}" --zone="${ZONE}" "${node}" --command "${cmd}"
}
```
?

**Comment by thockin** [comment]
I did that, but it will retry even if the command being run fails.  PR coming

**Comment by thockin** [comment]
running e2e first :)

**Comment by satnam6502** [comment]
Or "Russian Roulette" as I like to call it.

**Comment by brendandburns** [comment]
Closing this, since it appears to be fixed w/ #3962, no services.sh failures in the last N days, except ones due to etcd startup failure (which was fixed in a different PR)
